Introduce `CardFormView`

ccen-stripe opened the pull request:
# Summary
Introduce `CardFormView`, corresponds to iOS [STPCardFormView](https://github.com/stripe-ios/dashboard-ios/blob/87114bf3/Pods/Stripe/Stripe/STPCardFormView.swift#L1)

# Motivation
Introduce a modern UI for client to collect CardParams. Design details in [this](https://paper.dropbox.com/doc/API-Proposal-CardFormView--BIBfg2A05fkL1CbG7xeZ97A3Ag-w2duapoMsbJjSfPTDtxKe) internal doc

**note1**: At the moment `CardMultilineWidget` is reused with its style programmatically tweaked within `CardFormView`. As we gradually phase out `CardMultilineWidget`, it should be replaced with dedicated views for card details in order for `CardFormView` to support more flexible styles and customizations. 

**note2**: dedicated styles and colors are defined for `CardFormView`, but drawables for different cardBrand is reused from paymentsheet assets.

**note3**: `CardFormView` only returns valid `cardParams` only when _card details_, _country_ and _postal code_ are all valid. This is different from `CardMultilineWidget`, where _country_ and _postal code_ are optional

# Testing
<!-- How was the code tested? Be as specific as possible. -->
- [x] Added tests
- [ ] Modified tests
- [x] Manually verified

# Screen recordings

create 4 recordings for standard/borderless style with and without background color override

|            | with background override                                                                   | no background override                                                                     |
|------------|--------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------|
| Standard   | [link](https://drive.google.com/file/d/184FT9qrinT3LjuN_Ug-MXLzxdArTSdVy/view?usp=sharing) | [link](https://drive.google.com/file/d/1MvSxVBoG2Ej_mq1ev_VkXr73UhMkXugI/view?usp=sharing) |
| Borderless | [link](https://drive.google.com/file/d/1rXR86LRzQD5uZFW4VgySeHumL9RPpwPs/view?usp=sharing) | [link](https://drive.google.com/file/d/15PkUu-rNTWw4-rlZpikn1fpmd8kDy9oU/view?usp=sharing) |

mshafrir-stripe commented on stripe/res/values-night/colors.xml:
```diff
@@ -27,4 +27,10 @@
     <color name="stripe_paymentsheet_googlepay_divider_text">#99FFFFFF</color>
+
+    <!-- CardFormView -->
+    <color name="stripe_card_form_view_textinput_color">#DEFFFFFF</color>
```
nit: can we make these either all uppercase or all lowercase?

mshafrir-stripe commented on stripe/res/layout/card_form_view.xml:
```diff
@@ -0,0 +1,74 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Can we call this file `stripe_card_form_view.xml`?

mshafrir-stripe commented on stripe/src/test/java/com/stripe/android/view/CardFormViewTest.kt:
```diff
@@ -0,0 +1,280 @@
+        activityScenarioFactory.createView {
+            CardFormView(
+                context = it,
+                attrs = Robolectric.buildAttributeSet()
+                    .addAttribute(R.attr.cardFormStyle, "1")
+                    .build()
```
Cool, I didn't know about this!

mshafrir-stripe commented on stripe/src/main/java/com/stripe/android/view/CardFormView.kt:
```diff
@@ -0,0 +1,419 @@
+        ) {
+            formBackgroundColorStateList =
+                getColorStateList(R.styleable.StripeCardFormView_formBackgroundColorStateList)
+            style = Style.values()[getInt(R.styleable.StripeCardFormView_cardFormStyle, 0)]
```
Slight FUD about the enum value order being significant. Maybe update Style to include a `val attrValue: Int` property so we don't have to worry about a future order change breaking the integration?

```
enum class Style(
  internal val attrValue: Int
) {
  Standard(0),
  Borderless(1)
}
```

mshafrir-stripe commented on stripe/src/main/java/com/stripe/android/view/CardFormView.kt:
```diff
@@ -0,0 +1,419 @@
+        cardMultilineWidget.cardBrandIconSupplier =
+            CardMultilineWidget.CardBrandIconSupplier { cardBrand ->
+                CardMultilineWidget.CardBrandIcon(
+                    when (cardBrand) {
+                        CardBrand.Visa -> R.drawable.stripe_ic_paymentsheet_card_visa
+                        CardBrand.AmericanExpress -> R.drawable.stripe_ic_paymentsheet_card_amex
+                        CardBrand.Discover -> R.drawable.stripe_ic_paymentsheet_card_discover
+                        CardBrand.JCB -> R.drawable.stripe_ic_paymentsheet_card_jcb
+                        CardBrand.DinersClub -> R.drawable.stripe_ic_paymentsheet_card_dinersclub
+                        CardBrand.MasterCard -> R.drawable.stripe_ic_paymentsheet_card_mastercard
+                        CardBrand.UnionPay -> R.drawable.stripe_ic_paymentsheet_card_unionpay
+                        CardBrand.Unknown -> R.drawable.stripe_ic_paymentsheet_card_unknown
+                    }
```
I believe we have this logic in a few places. Can we move it out to a separate class?

ccen-stripe replied:
Sure, will do it in a separate ticket as this PR is getting pretty big

ccen-stripe commented:
updated with the following changes a bit after sync with @csabol-stripe 

1. move errors field outside of the card(see updated recording)
2. added a `CardValidCallback` and `setCardValidCallback` to notify if the current cardparams is ready
   - Note there is a bug on the callback triggered on card number and expiry fields - they get triggered twice because they [setText](https://github.com/stripe/stripe-android/blob/0552574078e467007fc7d2922bcf3ce41c7f6851/stripe/src/main/java/com/stripe/android/view/CardNumberEditText.kt#L382) again `afterTextChanged` to format the field. Will fix this in a separate PR

mshafrir-stripe commented on stripe/src/main/java/com/stripe/android/view/CardFormView.kt:
```diff
@@ -0,0 +1,419 @@
+ *
+ * To access the [CardParams], see details in [cardParams] property.
+ */
+class CardFormView @JvmOverloads constructor(
```
Flagging this

ccen-stripe replied:
done

mshafrir-stripe commented on stripe/src/main/java/com/stripe/android/view/CardFormView.kt:
```diff
@@ -0,0 +1,450 @@
+            }
+
+            // validation OK, clear any error view
+            setErrorMessageAndToggleVisibility(null)
```
maybe just `updateErrorsView`?

mshafrir-stripe commented on stripe/src/main/java/com/stripe/android/view/CardFormView.kt:
```diff
@@ -0,0 +1,450 @@
+/**
+ * A view to collect credit card information and provide [CardParams] for API invocation.
+ *
+ * Use [R.styleable.StripeCardFormView_cardFormStyle] to toggle style between [Style.Standard] and [Style.Borderless],
+ * Use [R.styleable.StripeCardFormView_formBackgroundColorStateList] to change the card's background color in enable and disabled state.
+ *
+ * To access the [CardParams], see details in [cardParams] property.
+ * To get notified if the current card params are valid, set a [CardValidCallback] object with [setCardValidCallback].
```
Thanks for adding this.

mshafrir-stripe commented on stripe/src/main/java/com/stripe/android/view/CardFormView.kt:
```diff
@@ -0,0 +1,450 @@
+import com.stripe.android.view.CardValidCallback.Fields
+
+/**
+ * A view to collect credit card information and provide [CardParams] for API invocation.
```
Can we add a comment about how the view takes the customer's country and adjusts the postal code view based on their selection?

mshafrir-stripe commented on stripe/src/main/java/com/stripe/android/view/CardFormView.kt:
```diff
@@ -0,0 +1,450 @@
+
+    private val cardContainer = viewBinding.cardMultilineWidgetContainer
+
+    @VisibleForTesting
+    internal val cardMultilineWidget = viewBinding.cardMultilineWidget
+
+    @VisibleForTesting
+    internal val countryPostalDivider = viewBinding.countryPostalDivider
+
+    @VisibleForTesting
+    internal val postalCodeContainer = viewBinding.postalCodeContainer
+
+    @VisibleForTesting
+    internal val errors = viewBinding.errors
+
+    @VisibleForTesting
+    internal val postalCodeView = viewBinding.postalCode
+
+    @VisibleForTesting
+    internal val countryLayout = viewBinding.countryLayout
+
```
Instead of making these views `internal` when they are only needed for testing, you can do something like. I really like this pattern.

```
@Test
fun `when locale is US then country should be US and postal config should be US`() {
    val viewBinding = StripeCardFormViewBinding.bind(standardCardFormView)
    assertThat(viewBinding.postalCode.text).isEqualTo("something")
}
```

ccen-stripe replied:
neat!

ccen-stripe commented on stripe/res/values/colors.xml:
```diff
@@ -48,4 +48,12 @@
+    <color name="stripe_card_form_view_textinput_color">#e61a1a1a</color>
+    <color name="stripe_card_form_view_background_default">@android:color/transparent</color>
+    <color name="stripe_card_form_view_background_disabled">#f8f8f8</color>
+    <color name="stripe_card_form_view_form_error">#cf6679</color>
```
hmm I copied this color from stripe_paymentsheet_form_error, which doesn't have night color either, should we keep them in sync?

mshafrir-stripe replied:
I'll defer to @michelleb-stripe

ccen-stripe replied:
discussed with michelle, looks like we're applying incorrect color palette for payment sheet from the [design](https://www.figma.com/file/2b9r3CJbyeVAmKi1VHV2h9/MC?node-id=1511%3A16122) 
Will change the color in a separate PR

mshafrir-stripe commented on stripe/src/main/java/com/stripe/android/view/CardValidCallback.kt:
```diff
@@ -15,6 +15,7 @@ fun interface CardValidCallback {
         Expiry,
-        Cvc
+        Cvc,
+        Zip
```
Postal is more generic

mshafrir-stripe commented on stripe/res/values/attrs.xml:
```diff
@@ -21,4 +21,12 @@
+            <enum name="STANDARD" value="0"/>
+            <enum name="BORDERLESS" value="1"/>
+        </attr>
+        <attr name="formBackgroundColorStateList" format="reference" />
```
Flagging

mshafrir-stripe commented on stripe/src/main/java/com/stripe/android/view/CardMultilineWidget.kt:
```diff
@@ -80,7 +80,7 @@ class CardMultilineWidget @JvmOverloads constructor(
         }
     }
-    private val invalidFields: Set<CardValidCallback.Fields>
+    val invalidFields: Set<CardValidCallback.Fields>
```
Does this need to be public?

ccen-stripe replied:
no, can be internal, changed

mshafrir-stripe commented on stripe/src/main/java/com/stripe/android/view/CountryTextInputLayout.kt:
```diff
@@ -125,6 +125,12 @@ internal class CountryTextInputLayout @JvmOverloads constructor(
 
+    override fun setEnabled(enabled: Boolean) {
+        super.setEnabled(enabled)
+        // setEnabled is called in super constructor, the empty check is required
+        countryAutocomplete?.isEnabled = enabled
```
Is there anyway to avoid this?

ccen-stripe replied:
Hmm I don't like it either, since `countryAutocomplete` can only be initialize after super constructor call and `setEnabled` is called in super constructor, the `val` non-null contract breaks here.
It's possible to define it as a `var countryAutocomplete: AutoCompleteTextView?`, but that would make all of its accessors have to use `.?`, which also isn't ideal.

Finally came up with the solution to create a separate `setEnabledWithChildView` function, `CardFormView` now calls this one instead of the default one so that we don't break super constructor. I guess this is not ideal either but less ugly ☹️

mshafrir-stripe commented on stripe/src/main/java/com/stripe/android/view/CountryTextInputLayout.kt:
```diff
@@ -125,10 +125,11 @@ internal class CountryTextInputLayout @JvmOverloads constructor(
+    internal fun setEnabledWithChildView(enabled: Boolean) {
+        isEnabled = enabled
+        countryAutocomplete.isEnabled = enabled
     }
```
Could we use `doOnNextLayout` to defer?

ccen-stripe replied:
nice catch!

mshafrir-stripe commented on stripe/res/values/attrs.xml:
```diff
@@ -21,4 +21,12 @@
         <attr name="countryItemLayout" format="reference" />
     </declare-styleable>
+
+    <declare-styleable name="StripeCardFormView">
+        <attr name="cardFormStyle" format="enum">
+            <enum name="standard" value="0"/>
+            <enum name="borderless" value="1"/>
+        </attr>
+        <attr name="formBackgroundColorStateList" format="reference" />
```
Can we remove the `form` prefix from `formBackgroundColorStateList`? Can be in a follow-up PR.
